### PR TITLE
Api2

### PR DESCRIPTION
**Pull request by Hellowlol**
Inital pr for the api. Its far from ready but it already has support for more commands then the old.

The idea is to reuse as much existing code as we can.

### CODE REVIEW

**Comment by Hellowlol** [comment]
I added a sql endpoint, while it gives great power/flexibiliy querying the db for the user but they can pretty much fuckup the entire db. Thoughts?

**Comment by drzoidberg33** [comment]
@Hellowlol How does the SQL endpoint work?

**Comment by Hellowlol** [comment]
Its very simple, it just takes a string and pass that do the db... :P So a user can do stuff like `DROP TABLE table_name;` if they want to.

https://github.com/drzoidberg33/plexpy/pull/377/files#diff-21dd9d3ce8c387b35ba33b0461068093R306

**Comment by drzoidberg33** [comment]
Ye, that sounds like a terrible idea :laughing:

**Comment by Hellowlol** [comment]
Hehe, I agree. But I do like the flexibility. How about I add it to the config so it has be enabled? (Will ofc be disabled by default.

**Comment by JonnyWong16** [comment]
Just a thought. It could also help with fixing other people's databases. Instead of instructions telling them to get SQLiteBrowser or something, just visit endpoint with the string. 

It's still dangerous though. Someone is going to come to us saying "I enabled this setting but I didn't know what it does. Now, I screwed up my database." Might not be worth the trouble.

**Comment by Hellowlol** [comment]
Yeah, thats why i wanted to add it to the config, it wouldnt even be a option in the ui to change it. its up to you guys, if you dont want it ill keep adding it in my own fork :)

Ill also add backup of the db. I will check if a backup was done today, if it wasnt it would backup the db before firing off the sql.

**Comment by Hellowlol** [comment]
What size is a average db? I'm just thinking about how many backups to keep

**Comment by JonnyWong16** [comment]
I would say keep at least 3 on a 12 or 24 hour schedule?

Backups:
- One is none.
- Two is one.
- Three is two.

**Comment by Hellowlol** [comment]
@drzoidberg33 @JonnyWong16 I think its ready, anybody got time to review the pr?

**Comment by JonnyWong16** [comment]
I'm on holiday until the new year starting tomorrow.

**Comment by Hellowlol** [comment]
https://github.com/drzoidberg33/plexpy/issues/376

**Comment by JonnyWong16** [comment]
@Hellowlol @wcomartin https://github.com/wcomartin/plexpy/commit/693869d5972a0d37a08b8c50585f523e98824145

Maybe you guys can merge these two together?

**Comment by Hellowlol** [comment]
@wcomartin I suggest you test the pr i have opened, while it isnt visable from the diff, alot of stuff is added to the api. I think i still need to pull your changes regarding gethistory since i have only copied the one that the ui uses. Use the docs_md command to see the api docs.

I dont see any benefit from splitting the api to be honest.

**Comment by wcomartin** [comment]
@Hellowlol I like what you've done here, unfortunately nothing i did would help, wish i'd known about this earlier, I'm going to fork your branch and see if i can help out though

also sorry i haven't responded earlier, i've been busy over the holidays

**Comment by Hellowlol** [comment]
@drzoidberg33 @JonnyWong16 anybody got time to review this pr?

**Comment by JonnyWong16** [comment]
I want to merge #373, #263, and #419 first.

**Comment by Hellowlol** [comment]
Ok :)

File: PlexPy.py [hunk 1]
```diff
@@ -49,6 +49,12 @@ def main():
 
     plexpy.PROG_DIR = os.path.dirname(plexpy.FULL_PATH)
     plexpy.ARGS = sys.argv[1:]
+    plexpy.BACKUP_DIR = os.path.join(plexpy.PROG_DIR, 'backups')
+
+    try:
```
Comment (JonnyWong16) [inline PlexPy.py]: This should be moved to the same place in the code where the other directories are created.
https://github.com/Hellowlol/plexpy/blob/master/plexpy/__init__.py#L128

File: plexpy/api2.py [hunk 2]
```diff
@@ -0,0 +1,541 @@
...
+            self._api_msg = 'API key not generated correctly'
+
+        elif 'apikey' not in kwargs:
+            self._api_msg = 'Parameter apikey is required'
+
+        elif kwargs.get('apikey', '') != plexpy.CONFIG.API_KEY:
+            self._api_msg = 'Invalid apikey'
+
+        elif 'cmd' not in kwargs:
+            self._api_msg = 'Parameter %s required. possible commands are: %s' % ', '.join(self._api_valid_methods)
```
Comment (JonnyWong16) [inline plexpy/api2.py]: Missing a string substitution here.

**Comment by Hellowlol** [comment]
Ok, i have rebased. Im sure i have broken some things, please test :P

**Comment by Hellowlol** [comment]
@drzoidberg33 @JonnyWong16 Got time to review this pr? Im getting tired of rebaseing it :D

**Comment by Hellowlol** [comment]
Zomg. Here we go again :D

**Comment by JonnyWong16** [comment]
Ah, sorry.

This one gets next priority.

**Comment by Hellowlol** [comment]
Rebased again fixing your remarks.

File: plexpy/api2.py [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+            self._api_msg = 'API key not generated correctly'
+
+        elif 'apikey' not in kwargs:
+            self._api_msg = 'Parameter apikey is required'
+
+        elif kwargs.get('apikey', '') != plexpy.CONFIG.API_KEY:
+            self._api_msg = 'Invalid apikey'
+
+        elif 'cmd' not in kwargs:
+            self._api_msg = 'Parameter cmd required. possible commands are: %s' % ', '.join(self._api_valid_methods)
```
Comment (JonnyWong16) [inline plexpy/api2.py]: `Parameter cmd is required. Possible commands are:`

File: plexpy/api2.py [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+
+        elif 'cmd' in kwargs and kwargs.get('cmd') not in self._api_valid_methods:
+            self._api_msg = 'Unknown command, %s possible commands are: %s' % (kwargs.get('cmd', ''), ', '.join(self._api_valid_methods))
```
Comment (JonnyWong16) [inline plexpy/api2.py]: `Unknown command: %s. Possible commands are:`

File: plexpy/api2.py [hunk 5]
```diff
@@ -0,0 +1,544 @@
...
+
+            Returns:
+                    string/dict: all settings as dict or one setting if key is provided
+        """
+
+        interface_dir = os.path.join(plexpy.PROG_DIR, 'data/interfaces/')
+        interface_list = [name for name in os.listdir(interface_dir) if
+                          os.path.isdir(os.path.join(interface_dir, name))]
+
+        config = {
```
Comment (JonnyWong16) [inline plexpy/api2.py]: Anyway to link this with the config from `webserver.py` so we don't have to change it in two places?

File: plexpy/webserve.py [hunk 6]
```diff
@@ -1276,6 +1318,7 @@ def get_notification_agent_triggers(self, config_id, **kwargs):
                               data=this_agent)
 
     @cherrypy.expose
+    @addtoapi()
```
Comment (JonnyWong16) [inline plexpy/webserve.py]: `@addtoapi('send_notification')` or `@addtoapi('notify')` or something else?

File: plexpy/webserve.py [hunk 7]
```diff
@@ -1580,11 +1633,8 @@ def get_search_results_children(self, query, media_type=None, season_index=None,
             logger.warn(u"Unable to retrieve data for get_search_results_children.")
             return serve_template(templatename="info_search_results_list.html", data=None, title="Search Result List")
 
-
-    
-    ##### Update Metadata #####
-
     @cherrypy.expose
+    @addtoapi()
```
Comment (JonnyWong16) [inline plexpy/webserve.py]: Not this one.

File: plexpy/webserve.py [hunk 8]
```diff
@@ -101,6 +102,7 @@ def welcome(self, **kwargs):
             return serve_template(templatename="welcome.html", title="Welcome", config=config)
 
     @cherrypy.expose
+    @addtoapi()
     def discover(self, token=''):
```
Comment (JonnyWong16) [inline plexpy/webserve.py]: Needs to check for a token before writing to the config. I accidentally wiped my token and couldn't figure out why stuff stopped working.

File: plexpy/webserve.py [hunk 9]
```diff
@@ -1709,6 +1782,7 @@ def get_recently_added_json(self, count='0', **kwargs):
             logger.warn(u"Unable to retrieve data for get_recently_added_json.")
 
     @cherrypy.expose
+    @addtoapi('get_episode_list')
```
Comment (JonnyWong16) [inline plexpy/webserve.py]: `get_episode_list_json` actually doesn't do anything. It can be removed completely.

File: plexpy/webserve.py [hunk 10]
```diff
@@ -242,6 +245,7 @@ def libraries(self):
         return serve_template(templatename="libraries.html", title="Libraries", config=config)
 
     @cherrypy.expose
+    @addtoapi()
     def get_library_list(self, **kwargs):
```
Comment (JonnyWong16) [inline plexpy/webserve.py:269]: All the datatable calls won't work with the API without providing the `json_data`.

File: plexpy/webserve.py [hunk 11]
```diff
@@ -251,6 +255,20 @@ def get_library_list(self, **kwargs):
...
+
+        if result:
+            cherrypy.response.headers['Content-type'] = 'application/json'
+            return json.dumps(result)
+        else:
+            logger.warn(u"Unable to retrieve data for get_library_sections.")
+
+    @cherrypy.expose
+    @addtoapi()
     def refresh_libraries_list(self, **kwargs):
```
Comment (JonnyWong16) [inline plexpy/webserve.py:304]: Commands which don't return anything (such as `refresh_libraries_list`) return error in the API.

```
{"response": {"message": null, "data": {}, "result": "error"}}
```

We should probably return something here.

File: plexpy/webserve.py [hunk 12]
```diff
@@ -1685,6 +1747,7 @@ def get_metadata_json(self, rating_key='', **kwargs):
             logger.warn(u"Unable to retrieve data for get_metadata_json.")
 
     @cherrypy.expose
+    @addtoapi('get_metadata')
```
Comment (JonnyWong16) [inline plexpy/webserve.py:1792]: Do we need `get_metadata_json` and `get_metadata_xml`?

File: plexpy/webserve.py [hunk 13]
```diff
@@ -1781,6 +1860,7 @@ def get_servers(self, **kwargs):
             logger.warn(u"Unable to retrieve data for get_servers.")
 
     @cherrypy.expose
+    @addtoapi()
     def get_servers_info(self, **kwargs):
```
Comment (JonnyWong16) [inline plexpy/webserve.py:1910]: Add `get_server_friendly_name` to API below?

File: plexpy/webserve.py [hunk 14]
```diff
@@ -1854,8 +1937,18 @@ def get_full_users_list(self, **kwargs):
         else:
             logger.warn(u"Unable to retrieve data for get_full_users_list.")
 
-    @cherrypy.expose
```
Comment (JonnyWong16) [inline plexpy/webserve.py:1868]: Oops?

**Comment by JonnyWong16** [comment]
APIv2 seems to work, but I'm finding it very messy. It's all inconsistent, but that's mostly because we're just adding a "random" bunch of existing functions.

We need docs for everything.

File: plexpy/api2.py [hunk 15]
```diff
@@ -0,0 +1,544 @@
...
+                                </response>
+                          ''' % e
+
+        return out
+
+    def _api_run(self, *args, **kwargs):
+        """ handles the stuff from the handler """
+
+        result = {}
+        logger.debug('Original kwargs was %s' % kwargs)
```
Comment (JonnyWong16) [inline plexpy/api2.py]: Add `APIv2` to all loggers? Makes it easier to identify and filter.

```
logger.debug(u"PlexPy APIv2 :: Original kwargs was %s" % kwargs)`
```

File: plexpy/webserve.py (hunk 12 above)
Comment (Hellowlol) [inline plexpy/webserve.py:1792]: No, initially I just decorated anything that didn't return a template :P

File: plexpy/api2.py (hunk 15 above)
Comment (Hellowlol) [inline plexpy/api2.py]: Yeah

File: plexpy/webserve.py (hunk 13 above)
Comment (Hellowlol) [inline plexpy/webserve.py:1910]: Why not? :S

**Comment by Hellowlol** [comment]
What i like about api2:
- Really easy to add new features
- I love that we can profile and debug the responses
- I like that we can choose the response type (dunno why ppl want xml but still..)
- Everything is documented. See api command docs_md
- Change docstrings to google style
- That we can rename function for consistency 

What i dont like:
- Cheezy handling of the responses since the decorator can return all sorts of crap.
- None responses, will always be a error, they need to return something or added to the api without the decorator
- The api docs as are good as our codes docs.... 
- The ghetto parser im using for the docs strings

There is probably other stuff that i like/dont like that i have forgotten :P
I need to remove some that functions that dont return a "clean" response. Like {"message": ""} since it dont fit with api response.

File: plexpy/api2.py (hunk 5 above)
Comment (Hellowlol) [inline plexpy/api2.py]: Yeah.

**Comment by Hellowlol** [comment]
I have lets most of the names as they are, you have any suggestion now how we can make it more consistent im all ears

**Comment by JonnyWong16** [comment]
Closed?

**Comment by Hellowlol** [comment]
Yes, tired of waiting and rebasing mate.

**Comment by JonnyWong16** [comment]
:(

I'll clean up and rebase later then.

**Comment by Hellowlol** [comment]
Alright, ill reopen and rebase it. Any thing you want change/removed?

**Comment by Hellowlol** [comment]
rebased

**Comment by JonnyWong16** [comment]
Merging so you don't have to keep rebasing. This also shouldn't affect anything else. I'll do some cleaning up later.

**Comment by Hellowlol** [comment]
Alright. Thanks

**Pull request merged**